internal/daemon/api_files.go

```diff
@@ -0,0 +1,576 @@
+
+	"github.com/canonical/pebble/internal/osutil"
+	"github.com/canonical/pebble/internal/osutil/sys"
+)
+
+const minBoundaryLength = 8
+
+func v1GetFiles(_ *Command, req *http.Request, _ *userState) Response {
+	query := req.URL.Query()
+	action := query.Get("action")
+	switch action {
+	case "read":
+		paths := query["path"]
+		if len(paths) == 0 {
+			return statusBadRequest("must specify one or more paths")
+		}
+		if req.Header.Get("Accept") != "multipart/form-data" {
+			return statusBadRequest(`must accept multipart/form-data`)
+		}
+		return readFilesResponse{paths: paths}
+	case "list":
+		path := query.Get("path")
+		if path == "" {
+			return statusBadRequest("must specify path")
+		}
+		pattern := query.Get("pattern")
+		itself := query.Get("itself")
+		if itself != "true" && itself != "false" && itself != "" {
+			return statusBadRequest(`itself parameter must be "true" or "false"`)
+		}
+		return listFilesResponse(path, pattern, itself == "true")
+	default:
+		return statusBadRequest("invalid action %q", action)
+	}
+}
+
+type fileResult struct {
+	Path  string       `json:"path"`
+	Error *errorResult `json:"error,omitempty"`
+}
+
+// Reading files
+
+// Custom Response implementation to serve the multipart.
+type readFilesResponse struct {
+	paths []string
+}
+
+func (r readFilesResponse) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	// Write HTTP status and headers. HTTP status is always OK because we don't
+	// know any better until we've read all the files.
+	mw := multipart.NewWriter(w)
+	header := w.Header()
+	header.Set("Content-Type", mw.FormDataContentType())
+	w.WriteHeader(http.StatusOK)
+
+	// Read each file's contents to multipart response.
+	result := make([]fileResult, len(r.paths))
+	for i, path := range r.paths {
+		err := readFile(path, mw)
+		result[i] = fileResult{
+			Path:  path,
+			Error: fileErrorToResult(err),
+		}
+	}
+
+	// At the end, write response metadata in JSON format.
+	mh := textproto.MIMEHeader{}
+	mh.Set("Content-Type", "application/json")
+	mh.Set("Content-Disposition", `form-data; name="response"`)
+	part, err := mw.CreatePart(mh)
+	if err != nil {
+		// Can't write metadata -- writing the error message on a new line is
+		// about the best we can do.
+		fmt.Fprint(w, "\n", err)
+		return
+	}
+	encoder := json.NewEncoder(part)
+	metadata := respJSON{
+		Type:       ResponseTypeSync,
+		Status:     http.StatusOK,
+		StatusText: http.StatusText(http.StatusOK),
+		Result:     result,
+	}
+	err = encoder.Encode(metadata)
+	if err != nil {
+		fmt.Fprint(w, "\n", err)
+		return
+	}
+
+	// Write the multipart trailer (last boundary marker).
+	err = mw.Close()
+	if err != nil {
+		fmt.Fprint(w, "\n", err)
+		return
+	}
+}
+
+func nonAbsolutePathError(path string) error {
+	return fmt.Errorf("paths must be absolute, got %q", path)
+}
+
+func readFile(path string, mw *multipart.Writer) error {
+	if !pathpkg.IsAbs(path) {
+		return nonAbsolutePathError(path)
+	}
+	if osutil.IsDir(path) {
```

Comment:
This should probably whitelist instead of blacklist (we don't want devices, sockets, etc). We should probably allow plain files only, but accept symlinks to plain files as well.

Reply:
Good call. Switched to use `os.Stat` with `info.Mode().IsRegular()`.